File: examples/jcdev/lib/jcdev/Renderer.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the forward declarations
#include "forward.h"
// external support
#include "externals.h"
// my declaration
#include "Renderer.h"


// metamethods
jcdev::Renderer::
Renderer()
{}


jcdev::Renderer::
~Renderer()
{}


// implementation details
auto
jcdev::Renderer::
render(const entry_type & entry) const -> record_type
{
    // get the page
    auto page = entry.page();
    // and the notes
    auto notes = entry.notes();

    // the joiner
    auto join = [] (string_t buffer, const string_t & line) {
                    return std::move(buffer) + '\n' + line;
                };
    // assemble the message
    string_t message = std::accumulate(next(page.begin()), page.end(), page[0], join);

    // make a record
    record_type record
        {
         // add the severity
         notes.at("severity"),
         // the application
         notes.at("application"),
         // the channel name
         notes.at("channel"),
         // the filename
         notes.at("filename"),
         // the line number
         notes.at("line"),
         // the function name
         notes.at("function"),
         // the message
         message
    };

    // and yield it
    return record;
}


// end of file

File: examples/jcdev/pkg/jcdev/Renderer.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


# the csv renderer or
class Renderer:
    """
    Renderer a journal entry as a CSV record
    """


    def render(self, entry):
        """
        Generate the CSV record
        """
        # get the page
        page = entry.page
        # and the notes
        notes = entry.notes

        # the severity
        yield notes["severity"]
        # the registered application name
        yield notes["application"]
        # the channel name
        yield notes["channel"]
        # the filename
        yield notes["filename"]
        # the line number
        yield notes["line"]
        # the function name
        yield notes["function"]

        # and finally, the page contents as a single string
        yield '\n'.join(page)

        # all done
        return


# end of file
